feat(cli): support single Ctrl+C to cancel streaming, preserving double Ctrl+C to exit

JayadityaGit opened the pull request:
### TL;DR

This PR improves the Gemini CLI experience by allowing a **single `Ctrl+C` press** to cancel an in-progress stream, matching common CLI expectations. The existing behavior of **double `Ctrl+C` to exit** the app is preserved.

---

### Summary of Changes

The main logic updates are in:

* `packages/cli/src/ui/App.tsx` – to detect a single `Ctrl+C` press during streaming.
* `packages/cli/src/ui/hooks/useGeminiStream.ts` – where the cancellation logic was extracted into a reusable `cancelRequest` function.

---

### Why This Matters

Previously, users needed to press the **`Esc` key** to cancel a long response — not a familiar pattern in most CLI tools. With this update, the CLI aligns with standard UX expectations:

* **Intuitive Interruption**: Pressing `Ctrl+C` once during streaming now cancels the stream immediately.
* **Unified Cancellation Logic**: Both `Esc` and `Ctrl+C` call a centralized `cancelRequest` function, reducing code duplication.
* **User Feedback**:

  * If you cancel during streaming: `Request cancelled.` is shown.
  * If no stream is active: `Press Ctrl+C again to exit.` is shown, just like before.

---

### How to Test It

1. **Start Gemini CLI.**
2. **Trigger a long streaming response** (e.g., ask it to write a novel about space-faring cats).
3. **Single `Ctrl+C` Cancellation**:

   * During the stream, press `Ctrl+C` once.
   * ✅ Expected: Stream cancels, you see `Request cancelled.`, and return to the input prompt.
4. **`Esc` Key Cancellation**:

   * Trigger another stream, press `Esc`.
   * ✅ Expected: Same behavior as `Ctrl+C`.
5. **Double `Ctrl+C` Exit**:

   * During a stream, press `Ctrl+C` twice quickly.
   * ✅ Expected: CLI exits.
6. **Idle `Ctrl+C` Handling**:

   * While idle (no stream), press `Ctrl+C` once.
   * ✅ Expected: Shows `Press Ctrl+C again to exit.` and waits. Press again within the window to exit.

---

### Testing Matrix

| Platform    | npm run | npx | Docker | Podman | Seatbelt |
| ----------- | ------- | --- | ------ | ------ | -------- |
| **Linux**   | ✅       | ❓   | ❓      | ❓      | ❓        |
| **macOS**   | ❓       | ❓   | ❓      | –      | –        |
| **Windows** | ❓       | ❓   | ❓      | –      | –        |

*(✅ = Tested; ❓ = Untested; – = Unsupported)*

---

### Linked Issue
#5837

JayadityaGit commented on packages/cli/src/ui/hooks/useGeminiStream.ts:
```diff
@@ -183,26 +183,36 @@ export const useGeminiStream = (
     return StreamingState.Idle;
   }, [isResponding, toolCalls]);
 
+  const cancelRequest = useCallback(() => {
+    if (turnCancelledRef.current) {
+      return;
+    }
+    turnCancelledRef.current = true;
+    abortControllerRef.current?.abort();
+    if (pendingHistoryItemRef.current) {
+      addItem(pendingHistoryItemRef.current, Date.now());
+    }
+    addItem(
+      {
+        type: MessageType.INFO,
+        text: 'Request cancelled.',
+      },
+      Date.now(),
+    );
+    setPendingHistoryItem(null);
+    onCancelSubmit();
+    setIsResponding(false);
+  }, [
+    addItem,
+    onCancelSubmit,
+    pendingHistoryItemRef,
+    setPendingHistoryItem,
+    setIsResponding,
+  ]);
```
Thanks !

JayadityaGit commented:
/gemini review

JayadityaGit commented:
I’d appreciate it if a human reviewer could take a look at this when possible. Thank you!

JayadityaGit commented on packages/cli/src/ui/hooks/useGeminiStream.ts:
```diff
@@ -183,26 +183,56 @@ export const useGeminiStream = (
+            : tool,
+        );
+        const updatedToolGroup: HistoryItemToolGroup = {
+          ...(pendingHistoryItemRef.current as HistoryItemToolGroup),
+          tools: updatedTools,
+        };
+        addItem(updatedToolGroup, Date.now());
+      } else {
         addItem(pendingHistoryItemRef.current, Date.now());
       }
-      addItem(
-        {
-          type: MessageType.INFO,
-          text: 'Request cancelled.',
-        },
-        Date.now(),
-      );
-      setPendingHistoryItem(null);
-      onCancelSubmit();
-      setIsResponding(false);
+    }
+    addItem(
+      {
+        type: MessageType.INFO,
+        text: 'Request cancelled.',
+      },
+      Date.now(),
+    );
+    setPendingHistoryItem(null);
+    onCancelSubmit();
+    setIsResponding(false);
+    setThought(null); // Also reset any active "thought" display.
+  }, [
+    addItem,
+    onCancelSubmit,
+    pendingHistoryItemRef,
+    setPendingHistoryItem,
+    setIsResponding,
+    setThought,
+  ]);
```
Thanks !

JayadityaGit commented:
/gemini review

jacob314 commented on packages/core/src/utils/environmentContext.test.ts:
```diff
@@ -77,7 +77,7 @@ describe('getEnvironmentContext', () => {
 
   beforeEach(() => {
     vi.useFakeTimers();
```
please revert changes to this file. they don't seem related.

jacob314 replied:
please revert

JayadityaGit commented:
/gemini review

JayadityaGit commented:
@jacob314 I’ve added the tests and updated the implementation so that a single Ctrl+C works the same as Esc. Thank you for your review.

JayadityaGit commented:
@jacob314 , The preflight check failed during the test:ci phase for the @google/gemini-cli package due to a JavaScript heap out of memory error. This occurred while running tests with coverage using Vitest. All 1388 tests in
  the dependent packages passed successfully.

  The failure seems related to resource constraints during the coverage reporting step rather than an actual test failure. Increasing the Node.js memory limit for the test process might resolve the issue. Consider
  running tests with NODE_OPTIONS="--max-old-space-size=4096" or investigating the coverage configuration.

jacob314 commented on eslint.config.js:
```diff
@@ -30,10 +30,7 @@ export default tseslint.config(
-      'packages/core/dist/**',
-      'packages/server/dist/**',
-      'packages/vscode-ide-companion/dist/**',
+      'packages/*/dist/**',
```
please revert

jacob314 commented on packages/cli/src/ui/App.test.tsx:
```diff
@@ -1119,4 +1119,253 @@ describe('App UI', () => {
+      await new Promise((resolve) => setTimeout(resolve, 10));
+
+      // Verify cancelRequest was called (which should handle tool call cancellation)
+      expect(mockCancelRequest).toHaveBeenCalled();
```
verify how the tool call is shown now that things were cancelled.

jacob314 commented on packages/cli/src/ui/App.test.tsx:
```diff
@@ -1119,4 +1119,253 @@ describe('App UI', () => {
+      expect(mockCancelRequest).toHaveBeenCalled();
+    });
+
+    it('should behave the same as Escape key for cancellation', async () => {
```
remove this test or fix so it makes more sense. it is a bit like two tests after each other.

jacob314 reviewed (commented):
approved after these comments are addressed. please be sure to revert the files that shouldn't have been changed. there was one file I commented on last round and another file added this round.

JayadityaGit commented:
Hi @jacob314, I’ve implemented the requested changes. Let me know if you have any further suggestions or improvements.

JayadityaGit commented:
@jacob314, I tried running vitest run --coverage but the job failed with a "JavaScript heap out of memory" error. Interestingly, the tests pass when run in isolation, so this might be related to memory usage during the full coverage run. Is there anything we can do to address this, such as optimizing the tests or increasing Node.js’s memory limit?

jacob314 commented:
Unfortunately there is a merge conflict. Please resolve and then I will merge.

JayadityaGit commented:
@jacob314 I’ve tried resolving the conflicts. Could you please review and confirm if everything looks correct? I’m still getting the hang of Git, so I’m a bit unsure.

JayadityaGit commented:
Hi @jacob314,

  Just wanted to provide an update on this PR.

  To resolve the merge conflicts with the main branch, I've rebased this feature branch (feat/ctrl-c-stream-stop) on top of the latest upstream/main.

  A conflict arose in packages/cli/src/ui/App.tsx during the rebase. We resolved it by integrating the cancelOngoingRequest() functionality from main into the newer, refactored key-handling logic in this branch.
  This preserves both the structural improvements and the request cancellation feature. The branch has been force-pushed, so this PR should now be up-to-date and conflict-free.

  As of now, this PR adds the intended functionality for Ctrl+C to cancel both ongoing streams and pending tool calls. I am still in the process of figuring out how to write the tests for this behavior.

  Thanks
